[openwrt-24.10] Improve support for Radxa's RK358x-based boards

1715173329 opened the pull request:
This is a backport of PR #19867.
Run tested on ROCK 5C Lite (RK3582).

Cc: @RadxaNaoki 

---

Improve support for the following Radxa RK358x-based boards:

* Radxa ROCK 5A
* Radxa ROCK 5B
* Radxa ROCK 5B+ (new)
* Radxa ROCK 5T (new)
* Radxa ROCK 5 ITX/ITX+ (new)
* Radxa ROCK 5C/5C Lite (new)
* Radxa E52C (new)

RadxaNaoki committed:
kernel: backport gated-fixed-clk driver

Backport gated-fixed-clk driver from Linux v6.13.

This is needed to fix a PCIe controller probe hang on the Radxa ROCK 5
ITX.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 9538c7fac04625000b64f3212348712498b443be)
[updated commit title]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
uboot-rockchip: prepare for Rockchip RK3582

Use "rk358x" for RK3588, RK3588S, and upcoming RK3582.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 9fa2a472af93db576507101de69f5e63c3a5ae00)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
uboot-rockchip: add support for Rockchip RK3582

Import patches from
 https://patchwork.ozlabs.org/project/uboot/list/?series=468562

Apply patch 3/3 partially due to conflicts.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 547626764bca4321b4aaf5108c7c416d9e472387)
[rebased upon 24.10 branch, added missing patches]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: backport dts changes for Rockchip RK358x

Backport dts changes up to Linux v6.17 for Rockchip RK358x.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 3adb68ef756f56c99f0305a807a6a365b028549b)
[rebased upon 24.10 branch, redo patches for kernel 6.6]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: backport dts changes for Radxa ROCK 5A

Backport dts changes up to Linux v6.17 for Radxa ROCK 5A.

Use power(green) LED instead of heartbeat(blue) LED.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 31a4923c4f46e91cf5a1229b1254afa735070ab2)
[rebased upon 24.10 branch, redo patches for kernel 6.6]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: backport dts changes for Radxa ROCK 5B/5B+/5T

Backport dts changes up to Linux v6.17 for Radxa ROCK 5B, 5B+, and 5T.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit c14178a71e8e051291d43620430ba3a29bcf1a76)
[rebased upon 24.10 branch, redo patches for kernel 6.6]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: backport dts changes for Radxa ROCK 5 ITX/ITX+

Backport dts changes up to Linux v6.17 for Radxa ROCK 5 ITX and ITX+.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 61acc92a71c8cb16580c771d2c436d558d4a2d44)
[added missing patch, redo patches for kernel 6.6]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: backport dts changes for Radxa ROCK 5C/5C Lite

Backport dts changes up to Linux v6.17 for Radxa ROCK 5C and 5C Lite.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 73d50b420e545d3beebcf00430600a1fbb3ac8b4)
[rebased upon 24.10 branch, redo patches for kernel 6.6]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: backport dts changes for Radxa E52C

Backport dts changes up to Linux v6.17 for Radxa E52C.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 6e361c76ee3752a6e3634bc5d2f057cdeb39540b)
[rebased upon 24.10 branch]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add HATs/M.2 E key module support for Radxa ROCK 5A

Add support for Radxa HATs/M.2 E key module.

- Radxa Dual 2.5G Router HAT (Realtek RTL8125BG/NVMe)
- Radxa Penta SATA HAT (JMicron JMB585)
- Radxa Wireless Module A8 (Realtek RTL8852BE)

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 3c23e9a32754f52f50cd0f66f94c32ea01e462a5)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add M.2 M/E key modules support for Radxa ROCK 5B

Add support for M.2 M/E key modules.

- NVMe
- Radxa Wireless Module A8 (Realtek RTL8852BE)

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 8aa83efe5755720494a9a23d051b97172427f451)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add support for Radxa ROCK 5B+

The Radxa ROCK 5B+ is a single board computer using the Rockchip
RK3588 SoC.

- https://radxa.com/products/rock5/5bp

Hardware
--------

- Quad Cortex-A76 and Quad Cortex-A55 CPU
- Mali-G610 MP4 GPU
- 6 TOPS NPU
- 4/8/16/24/32GB LPDDR5 RAM
- 16/32/64/128/256GB on-board eMMC (optional)
- 16MB SPI NOR flash
- microSD card slot
- Wi-Fi 6 (Realtek RTW8852BE)
- 2.5 Gigabit Ethernet port with PoE (additional PoE HAT required)
- USB 3.1 Gen1 Type-C HOST/OTG port
- 2x USB 3.1 Gen1 Type-A HOST ports
- 2x USB 2.0 Type-A HOST ports
- 2x M.2 M Key connectors with PCIe 3.0 x2
- M.2 B Key connector
- SIM card slot
- RTC battery connector
- PWM fan connector
- 20x2 pin header
- USB Type-C power port

Installation
------------

Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
eMMC using dd.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit d35c641ced9dd3cfe74b94aca26e21db0cc406de)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add support for Radxa ROCK 5T

The Radxa ROCK 5T is a single board computer using the Rockchip
RK3588(J) SoC.

- https://radxa.com/products/rock5/5t

Hardware
--------

- Quad Cortex-A76 and Quad Cortex-A55 CPU
- Mali-G610 MP4 GPU
- 6 TOPS NPU
- 4/8/16/24/32GB LPDDR5 RAM
- 16/32/64/128/256GB on-board eMMC (optional)
- 16MB SPI NOR flash
- microSD card slot
- Wi-Fi 6E (Intel AX210, limited support)
- 2x 2.5 Gigabit Ethernet ports with PoE (additional PoE module required)
- USB 3.1 Gen1 Type-C HOST/OTG port
- 2x USB 3.1 Gen1 Type-A HOST ports
- 2x USB 2.0 Type-A HOST ports
- 2x M.2 M Key connectors with PCIe 3.0 x2
- M.2 B Key connector
- SIM card slot
- RTC battery connector
- PWM fan connector
- 20x2 pin header
- 5525 12V DC jack

Installation
------------

Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
eMMC using dd.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 4a78af98766b252282d50bdc0de32cebad71944d)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add support for Radxa ROCK 5 ITX/ITX+

The Radxa ROCK 5 ITX(+) is a Mini-ITX form factor computer using the
Rockchip RK3588 SoC.

- https://radxa.com/products/rock5/5itx
- https://radxa.com/products/rock5/5itxp

Hardware
--------

- Quad Cortex-A76 and Quad Cortex-A55 CPU
- Mali-G610 MP4 GPU
- 6 TOPS NPU
- 4/8/16/24/32GB LPDDR5 RAM
- on-board eMMC
- 16MB SPI NOR flash
- microSD card slot
- 2x 2.5 Gigabit Ethernet ports with PoE (additional PoE module required)
- USB 3.1 Gen1 Type-C HOST/OTG port
- 4x USB 3.1 Gen1 Type-A HOST ports
- 2x USB 2.0 Type-A HOST ports
- M.2 M Key connector with PCIe 3.0 x2
- 4x SATA connectors (ITX only)
- 2nd M.2 M Key connector with PCIe 3.0 x2 (ITX+ only)
- M.2 E Key connector with PCIe 2.1 x1 and USB 2.0
- RTC battery socket for CR1220
- 4pin PWM fan connector
- Serial console pin header
- Front panel pin headers
- 24pin ATX power connector
- 5525 12V DC jack

Installation
------------

Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
eMMC using dd.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 083934521186d54ded432c16a91e5fcea73c8308)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add support for Radxa ROCK 5C/5C Lite

The Radxa ROCK 5C (Lite) is a single board computer using the Rockchip
RK3588S2 (RK3582) SoC.

- https://radxa.com/products/rock5/5c

Hardware
--------

- Quad (Dual) Cortex-A76 and Quad Cortex-A55 CPU
- Mali-G610 MP4 GPU (5C only)
- 6 (5) TOPS NPU
- 1/2/4/8/16/32GB LPDDR4X RAM
- eMMC/SPI NOR flash connector
- microSD card slot
- Wi-Fi 6 (AIC8800D80, not yet supported)
- Gigabit Ethernet port with PoE (additional PoE HAT required)
- USB 3.1 Gen1 Type-A HOST/OTG port
- USB 3.1 Gen1 Type-A HOST port
- 2x USB 2.0 Type-A HOST ports
- FPC connector with PCIe 2.1 x1
- PWM fan connector
- 20x2 pin header
- USB Type-C power port

Installation
------------

Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
eMMC using dd.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 071aa994cdd1f716f2f9e0f2e2728823c5c9b753)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki committed:
rockchip: add support for Radxa E52C

The Radxa E52C is a compact network computer using the Rockchip RK3582
SoC.

- https://radxa.com/products/network-computer/e52c

Hardware
--------

- Dual Cortex-A76 and Quad Cortex-A55 CPU
- 5 TOPS NPU
- 2/4/8GB LPDDR4 RAM
- 16/32/64GB on-board eMMC
- microSD card slot
- 2x 2.5 Gigabit Ethernet ports
- USB 3.1 Gen1 Type-A HOST/OTG port
- USB Type-C debug port
- USB Type-C power port

Installation
------------

Uncompress the OpenWrt sysupgrade and write it to a micro SD card or
eMMC using dd.

Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/19867
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit d16d2765bd4fba96935f2a5ad6f4e830da317956)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

1715173329 commented:
> I'd like to test your patch, but I'm currently busy with another task and expect to be able to test it in a few days.

Thank you!

1715173329 commented:
Yes, so I added some hacks[^1][^2][^3] to ImmortalWrt for known affected devices, but this way is already explicitly rejected by upstream kernel as they want us to handle the ifname in userspace (like udev).

static-device-names looks lovely, maybe it should be added into core repo.

[^1]: https://github.com/immortalwrt/immortalwrt/blob/master/target/linux/rockchip/patches-6.12/601-net-ethernet-r8169-add-devname-configuration-from-OF.patch
[^2]: https://github.com/immortalwrt/immortalwrt/blob/master/target/linux/rockchip/patches-6.12/602-net-ethernet-stmmac-add-devname-configuration-from-OF.patch
[^3]: https://github.com/immortalwrt/immortalwrt/blob/master/target/linux/rockchip/patches-6.12/610-arm64-rockchip-add-OF-node-for-eth.patch

RadxaNaoki commented:
I performed tests on the following boards and confirmed that the expected speeds are achieved with iperf3:

* Radxa ROCK 5A
* Radxa ROCK 5B
* Radxa ROCK 5B+
* Radxa ROCK 5T
* Radxa ROCK 5 ITX+
* Radxa ROCK 5C
* Radxa E52C

Basically, it looks good to me.

However, I observed an issue where eth0 and eth1 swap due to the PCIe bus probe order changing. I would like to improve this, if possible.

RadxaNaoki committed:
rockchip: make NIC name predictable for Radxa E52C/ROCK 5 ITX/ROCK 5T

The probe order for PCIe buses and devices is non-deterministic,
making the names eth0 and eth1 unpredictable (they may be swapped).

This patch fixes the names by referencing the device path using
`ucidef_set_network_device_path`.

The mapping between silkscreen labels on the board/case and OpenWrt
interface names is as follows:

- E52C
  LAN: lan
  WAN: wan

- ROCK 5 ITX
  RJ45 1: eth0
  RJ45 2: eth1

- ROCK 5T
  RJ45_1: eth0
  RJ45_2: eth1

For Radxa E52C, this breaks compatibility of the network config;
therefore, set DEVICE_COMPAT_VERSION to `1.1`.

Fixes: d16d2765bd ("rockchip: add support for Radxa E52C")
Fixes: 0839345211 ("rockchip: add support for Radxa ROCK 5 ITX/ITX+")
Fixes: 4a78af9876 ("rockchip: add support for Radxa ROCK 5T")
Link: https://github.com/openwrt/openwrt/issues/20202
Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/20608
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit 1f1db7543220b70610782bdaa56050466094c14a)
[rebased upon 24.10 branch]
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki commented:
@hauke When this PR is merged, please merge #20711 too.

1715173329 committed:
rockchip: fix assign IRQ SMP affinity for Radxa E52C

The ifname was renamed to `lan`/`wan` in commit 1f1db7543220
 ("rockchip: make NIC name predictable for Radxa E52C/ROCK 5 ITX/ROCK 5T"),
update accordingly.

Fixes: 1f1db7543220 ("rockchip: make NIC name predictable for Radxa E52C/ROCK 5 ITX/ROCK 5T")
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>
Link: https://github.com/openwrt/openwrt/pull/20722
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
(cherry picked from commit d720ad52543a5ead7e7e9a3f201bea3ae2619f75)
Signed-off-by: Tianling Shen <cnsztl@immortalwrt.org>

RadxaNaoki commented:
Oh I forgot to fix 40-net-smp-affinity in #20608

1715173329 commented:
I've created a PR for it #20722

RadxaNaoki committed:
rockchip: add `"compat_version": "1.1"` for Radxa E52C

This is something I missed in https://github.com/openwrt/openwrt/pull/20608

Fixes: 1f1db75432 ("rockchip: make NIC name predictable for Radxa E52C/ROCK 5 ITX/ROCK 5T")
Signed-off-by: FUKAUMI Naoki <naoki@radxa.com>
Link: https://github.com/openwrt/openwrt/pull/21533
Signed-off-by: Robert Marko <robimarko@gmail.com>
(cherry picked from commit 43039157babc7262180cdd56994732ca1ebbb6fd)
